files/gen_sys_stats.rb

```diff
@@ -0,0 +1,252 @@
+    #
+    # @param [string] sar_output
+    # @param [obj] time_stamp_obj Time object to use for generating the filename
+    #
+    # @return [string] json of sar output
+    def convert_sar_output_to_json(sar_output, time_stamp_obj)
+      header = get_header_fields(sar_output)
+      averages = get_averages_fields(sar_output)
+      hostkey = @hostname.gsub('.', '-')
+
+      dataset = {'time_stamp_obj' => time_stamp_obj.utc.iso8601, 'servers' => {}}
+      #combine the arrays into a hash and convert into json
+      metrics_data = Hash[header.zip(averages)]
+      dataset['servers'][hostkey] = {@metric_type => metrics_data}
+      json_dataset = JSON.pretty_generate(dataset)
+      return json_dataset
+    end
+
+    # Create the file and put the json data in it
+    #
+    # @author Randell Pelak
+    #
+    # @param [string] json_dataset data in json format to put in file
+    # @param [obj] time_stamp_obj Time object to use for generating the filename
+    #
+    # @return [void]
+    def create_file(json_dataset, time_stamp_obj)
+      filename = time_stamp_obj.utc.strftime('%Y%m%dT%H%M%SZ') + '.json'
+      dirname = "#{@metrics_dir}/system_#{@metric_type}/#{@hostname}"
+      file_path = "#{dirname}/#{filename}"
+      FileUtils.mkdir_p(dirname) unless File.directory?(dirname)
+      puts "Creating json file: #{file_path}" if @verbose
+      File.open(file_path, 'w') do |file|
+        file.write(json_dataset)
+      end
+    end
+
+
+    # Get the data and create the file
+    #
+    # @author Randell Pelak
+    #
+    # @return [void]
+    def go
+      output = run_sar
+      raise(_("sar output invalid")) unless sar_output_valid?(output)
+      time_stamp_obj = Time.now
+      json_data = convert_sar_output_to_json(output, time_stamp_obj)
+      create_file(json_data,time_stamp_obj)
+    end
+
+  end
+end
+
+if $PROGRAM_NAME == __FILE__
+
+  VALID_METRIC_TYPES = %w[cpu memory]
+  FILE_INTERVAL_DEFAULT = 60 * 5
+  POLLING_INTERVAL_DEFAULT = 1
+  METRIC_TYPE_DEFAULT = "cpu"
+  METRICS_DIR_DEFAULT = "/opt/puppetlabs/puppet-metrics-collector"
+
+  DESCRIPTION = <<-DESCRIPTION
+    This script is intended to be run on a puppet infrastructure node as part of the
+    puppet_metrics_collector module.  It will generate system statistics in the file format that
+    other puppet_metrics_collector stats follow.  It will poll the system for data at a given
+    interval, and then output the average to a file once per given file interval.
+  DESCRIPTION
+
+  DEFAULTS = <<-DEFAULTS
+    The following defaults values are used if the options are not specified:
+      * polling_interval (-p, --polling_interval): #{POLLING_INTERVAL_DEFAULT}
+      * file_interval (-f, --file_interval): #{FILE_INTERVAL_DEFAULT}
+      * metric_type (-t, --metric_type): #{METRIC_TYPE_DEFAULT}
+      * metrics_dir (-m, --metrics_dir): #{METRICS_DIR_DEFAULT}
+      * verbose (-v, --verbose): False
+  DEFAULTS
+
+  options = {}
+
+  OptionParser.new do |opts|
```

Comment:
This takes up many lines.  I am no expert here, but I am wondering if there is another library that could reduce the CLI boilerplate.
Puppet uses one called [trollop](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/command_line/trollop.rb) which has been renamed to [optimist](https://www.manageiq.org/optimist/).  I have also bumped into others using [thor](http://whatisthor.com/).

Reply:
I think it is kind of a path of least resistance at this point.  We have used it in several places and are at least familiar with it.  We could make a ticket to investigate others at some point...